### ISSUE TITLE
Make it possible to set advertise-address to the internal IP

### ISSUE THREAD

**Issue by rbjorklin** [syself/cluster-api-provider-hetzner#605]
/kind feature

**Describe the solution you'd like**
It would be great to be able to configure `advertise-address` some way so that one could easily firewall incoming traffic from the internet.

**Anything else you would like to add:**
If this is already possible it would be great to have it documented somewhere.

**Environment:**

- cluster-api-provider-hetzner version: 
- Kubernetes version: (use `kubectl version`)
- OS (e.g. from `/etc/os-release`):

**Comment by BlakeB415** [comment]
I found a decent workaround temporarily

When you have private networks enabled, we know that the first thing attached to it will be the API load balancer. This internal address will be ``10.0.0.2``

Then we can set
```
controlPlaneEndpoint:
  host: "10.0.0.2"
  port: 443
```
on the HetznerCluster CRD.

If you want to be able to connect externally, you would need to add an A record to a domain name.
Then add this to the KubeadmControlPlane CRD.
```
  kubeadmConfigSpec:
    clusterConfiguration:
      apiServer:
        certSANs:
          - kubernetes.example.com
```

You would need to make sure the CAPI services can access the cluster via the private addresses.

**Comment by rbjorklin** [comment]
I haven't worked out all the kinks yet but I've put together a little [example repo to create a NAT gateway](https://github.com/rbjorklin/hetzner-nat-gateway) which side-steps this problem quite nicely.

**Comment by chess-knight** [comment]
When https://github.com/canonical/cloud-init/issues/4263 is resolved, IMHO it should be somehow possible to set localAPIEndpoint.advertiseAddress via cloud-init jinja templating.

**Comment by guettli** [comment]
@rbjorklin is this issue a duplicate of https://github.com/syself/cluster-api-provider-hetzner/issues/762 ?

**Comment by simonostendorf** [comment]
> @rbjorklin is this issue a duplicate of #762 ?

@guettli I dont think that setting advertise-address (which is the initial feature request) has something to do with a pre-created cloud network. For me these are two different aspects. But I think that setting advertise-address is not a specific problem of CAPH, this should be discussed on CAPI level.

**Comment by guettli** [comment]
@simonostendorf thank you for your answer. Do you know if there is an issue or PR for that on capi level?

I guess we can't do here much at the moment. Is it ok to close the issue?

**Comment by rbjorklin** [comment]
I wouldn't mind closing this. For my own use-case I would much prefer running a cluster on public only IPv6 addresses and protect my cluster through strict firewall rules.

**Comment by tboerger** [comment]
If https://github.com/canonical/cloud-init/pull/6224 gets merged you should be able to reference the private IPs within the cloud-init config. With this change you should be able to use something like `{{ ds.meta_data.private-networks.network_name.ip }}` for `bind-address` or `advertise-client-urls` flags.

**Comment by simonostendorf** [comment]
> If https://github.com/canonical/cloud-init/pull/6224 gets merged you should be able to reference the private IPs within the cloud-init config. With this change you should be able to use something like `{{ ds.meta_data.private-networks.network_name.ip }}` for `bind-address` or `advertise-client-urls` flags.

I don't know if this works all the time because I have Hetzner nodes that need 2 minutes to attach the private IP, so cloud init is already finished at that point.

**Comment by tboerger** [comment]
> Do you know if the Metadata API already serves the information about the private network even if the interface is not fully attached?

The API serves this information as soon as a private network is attached to the VM, at least that's what I was told.

**Comment by tboerger** [comment]
There had been single cases where the attribute have been empty, but even if I ignore this there are still hard-coded values inside the deployments. That are still the public IPs if I grep for them within the /etc/kubernetes/manifests folder:

```
/etc/kubernetes/manifests/etcd.yaml:    kubeadm.kubernetes.io/etcd.advertise-client-urls: https://128.140.93.250:2379
/etc/kubernetes/manifests/etcd.yaml:    - --initial-cluster=mgmt-control-plane-tl2ls=https://128.140.93.250:2380
/etc/kubernetes/manifests/kube-apiserver.yaml:    kubeadm.kubernetes.io/kube-apiserver.advertise-address.endpoint: 128.140.93.250:6443
/etc/kubernetes/manifests/kube-apiserver.yaml:        host: 128.140.93.250
/etc/kubernetes/manifests/kube-apiserver.yaml:        host: 128.140.93.250
/etc/kubernetes/manifests/kube-apiserver.yaml:        host: 128.140.93.250
```

So saldy I'm still not able to do a proper private IP networking setup with CAPH as long as the Hetzner instance got a public IP and that sucks pretty hard.

**Comment by tboerger** [comment]
And what have you done? It with sed?

**Comment by tboerger** [comment]
Answering my own question: `sed -i "s/PRIVATE_IP/{{ ds.meta_data.private_networks.mgmt.ip }}/g" /run/kubeadm/kubeadm.yaml`

**Comment by tboerger** [comment]
Finally I got everything hacked together to have all components listening to private ips only. The only downside is that I'm provisioning by management cluster from local kind where I want to move the components to the bootstraped cluster. This won't start my 3 conducted control plane nodes as clusterapi on kind things that etcd is not healthy 🙈

**Comment by simonostendorf** [comment]
> Finally I got everything hacked together to have all components listening to private ips only. The only downside is that I'm provisioning by management cluster from local kind where I want to move the components to the bootstraped cluster. This won't start my 3 conducted control plane nodes as clusterapi on kind things that etcd is not healthy 🙈

Do you have an public API endpoint via LB? I think its checked via api because I dont have these problems

**Comment by tboerger** [comment]
Yes and the lb works for the Kubernetes api, but it's trying to check the etcd health which is not expressed via lb.

**Comment by simonostendorf** [comment]
> Yes and the lb works for the Kubernetes api, but it's trying to check the etcd health which is not expressed via lb.

I think cluster api checks the etcd health via the api lb, I dont think there is an ssh login or something else. Because I already used clusters that only had private node ips.

See control plane provider code how the etcd member health is checked: https://github.com/kubernetes-sigs/cluster-api/blob/main/controlplane/kubeadm/internal/workload_cluster_conditions.go#L129

**Comment by simonostendorf** [comment]
> I set `spec.kubeadmConfigSpec.initConfiguration.localAPIEndpoint.advertiseAddress` and `spec.kubeadmConfigSpec.joinConfiguration.controlPlane.localAPIEndpoint.advertiseAddress` to a temporary value in the `KubeadmControlPlane` resource and replaced it in the preKubeadmCommands with the private node ip and this works I guess.

Addition to this comment: I dont do this anymore since ControlPlaneKubeletLocalMode is in beta, IP addresses are working out of the box with 1.33 and standard hccm.
